nomad/state/events.go

```diff
@@ -0,0 +1,242 @@
+type NodeEvent struct {
+	Node *structs.Node
+}
+
+// NNodeDrainEvent is the Payload for a NodeDrain event. It contains
+// information related to the Node being drained as well as high level
+// information about the current allocations on the Node
+type NodeDrainEvent struct {
+	Node      *structs.Node
+	JobAllocs map[string]*JobDrainDetails
+}
+
+type NodeDrainAllocDetails struct {
+	ID      string
+	Migrate *structs.MigrateStrategy
+}
+
+type JobDrainDetails struct {
+	Type         string
+	AllocDetails map[string]NodeDrainAllocDetails
+}
+
+// GenericEventsFromChanges returns a set of events for a given set of
+// transaction changes. It currently ignores Delete operations.
+func GenericEventsFromChanges(tx ReadTxn, changes Changes) (*structs.Events, error) {
+	var eventType string
+	switch changes.MsgType {
+	case structs.NodeRegisterRequestType:
+		eventType = TypeNodeRegistration
+	case structs.UpsertNodeEventsType:
+		eventType = TypeNodeEvent
+	case structs.EvalUpdateRequestType:
+		eventType = TypeEvalUpdated
+	case structs.AllocClientUpdateRequestType:
+		eventType = TypeAllocUpdated
+	case structs.JobRegisterRequestType:
+		eventType = TypeJobRegistered
+	case structs.AllocUpdateRequestType:
+		eventType = TypeAllocUpdated
+	case structs.NodeUpdateStatusRequestType:
+		eventType = TypeNodeEvent
+	case structs.JobDeregisterRequestType:
+		eventType = TypeJobDeregistered
+	case structs.JobBatchDeregisterRequestType:
+		eventType = TypeJobBatchDeregistered
+	case structs.AllocUpdateDesiredTransitionRequestType:
+		eventType = TypeAllocUpdateDesiredStatus
+	case structs.NodeUpdateEligibilityRequestType:
+		eventType = TypeNodeDrain
+	case structs.BatchNodeUpdateDrainRequestType:
+		eventType = TypeNodeDrain
+	case structs.DeploymentStatusUpdateRequestType:
+		eventType = TypeDeploymentUpdate
+	case structs.DeploymentPromoteRequestType:
+		eventType = TypeDeploymentPromotion
+	case structs.DeploymentAllocHealthRequestType:
+		eventType = TypeDeploymentAllocHealth
+	case structs.ApplyPlanResultsRequestType:
+		eventType = TypePlanResult
+	default:
+		// unknown request type
+		return nil, nil
+	}
+
+	var events []structs.Event
+	for _, change := range changes.Changes {
+		switch change.Table {
+		case "evals":
+			if change.Deleted() {
+				return nil, nil
+			}
+			after, ok := change.After.(*structs.Evaluation)
+			if !ok {
+				return nil, fmt.Errorf("transaction change was not an Evaluation")
+			}
+
+			event := structs.Event{
+				Topic:     TopicEval,
+				Type:      eventType,
+				Index:     changes.Index,
+				Key:       after.ID,
+				Namespace: after.Namespace,
+				Payload: &EvalEvent{
+					Eval: after,
+				},
+			}
+
+			events = append(events, event)
+
+		case "allocs":
+			if change.Deleted() {
+				return nil, nil
+			}
+			after, ok := change.After.(*structs.Allocation)
+			if !ok {
+				return nil, fmt.Errorf("transaction change was not an Allocation")
+			}
+
+			alloc := after.Copy()
```

Comment:
actually we do need a copy here, structs _shouldn't_ be mutated in place, since the updates to the pointer won't be saved across servers, but it will definitely impact others using that alloc pointer